# Task -02
## Data cleaning and Explorartory Data Anaylisis

###  Data Cleaning

####  Import  libraries

In [468]:
#type:ignore 
import pandas as pd                    # for data manipulation and analysis
import numpy as np                     # for numerical operations and array manipulation
import seaborn as sns                  # for statistical data visualization
import matplotlib.pyplot as plt        # for creating visualizations

#### Load the Titanic dataset

In [469]:
# Load the Titanic dataset
Titanic = pd.read_csv("Titanic _Dataset.csv")

#### Understanding the dataset

In [470]:
#type:ignore
 
# Check the basic info of the dataset
Titanic.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


#### Code Description
The `Titanic.info()` function provides a summary of the dataset, including:
- The number of entries
- Column names
- Data types
- Non-null values

#### Titanic Dataset Column Descriptions
---
- **PassengerId**: A unique identifier for each passenger.
- **Survived**: Indicates whether the passenger survived (1) or did not survive (0).
- **Pclass**: The passenger's ticket class (1 = first class, 2 = second class, 3 = third class).
- **Name**: The name of the passenger.
- **Sex**: The gender of the passenger (male or female).
- **Age**: The age of the passenger in years.
- **SibSp**: The number of siblings or spouses the passenger had on board.
- **Parch**: The number of parents or children the passenger had on board.
- **Ticket**: The ticket number of the passenger.
- **Fare**: The fare paid by the passenger for the ticket, in British pounds. This reflects the travel class and amenities, influencing survival chances; higher fares often indicated first-class accommodations.
- **Cabin**: The cabin number assigned to the passenger, indicating their location on the ship. Cabin location played a role in access to lifeboats during the evacuation, affecting survival rates.
- **Embarked**: The port of embarkation (C = Cherbourg, Q = Queenstown, S = Southampton)
---

#### Preview of the Titanic Dataset

In [471]:
Titanic.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

#### Code Description
---
- The `Titanic.head()`function displays the first five rows of the dataset, giving a quick overview of the data structure and the initial values in each column.
---

#####  Explanation of the Steps
---
- Checking for duplicates is done first to ensure the dataset's integrity before any further processing, as duplicates can skew analysis results.
- Next, verifying data types ensures that each column is correctly formatted for the intended operations, facilitating accurate calculations and analyses.
- Finally, filling missing values addresses gaps in the data to maintain completeness and reliability in the dataset, ensuring more robust analysis outcomes.
---


#### Check for Duplicates

In [472]:
# Check for duplicates
duplicates = Titanic.duplicated().sum()
print(f"Number of duplicate entries: {duplicates}")

Number of duplicate entries: 0


##### Code Description
---
- `Titanic.duplicated().sum()` This code checks for duplicate entries in the Titanic DataFrame by identifying rows that are identical and counts them, outputting the total number of duplicates.
---

In [473]:
# Check data types
print("Data types")
Titanic.dtypes

Data types


PassengerId      int64
Survived         int64
Pclass           int64
Name            object
Sex             object
Age            float64
SibSp            int64
Parch            int64
Ticket          object
Fare           float64
Cabin           object
Embarked        object
dtype: object

#####  Transforming to convenient data types

In [474]:
# Specifying the columns to convert
cols = ['Pclass','Name', 'Sex', 'Ticket', 'Cabin', 'Embarked']

# Converting the specified columns to categorical types
for col in cols:
    Titanic[col] = Titanic[col].astype('category')

# Check the data types to confirm the changes
print("\nData types after conversion:")
print(Titanic.dtypes)




Data types after conversion:
PassengerId       int64
Survived          int64
Pclass         category
Name           category
Sex            category
Age             float64
SibSp             int64
Parch             int64
Ticket         category
Fare            float64
Cabin          category
Embarked       category
dtype: object


##### Code Description
---
This code snippet converts specified columns in the Titanic DataFrame to categorical data types. Categorical types are more memory-efficient and enhance the performance of data analysis tasks. The columns being converted are:

-**Name**: The full name of the passenger.

-**Sex**: The gender of the passenger (male or female).

-**Ticket**: The ticket number of the passenger.

-**Cabin**: The cabin number assigned to the passenger.

-**Embarked**: The port of embarkation (C = Cherbourg, Q = Queenstown, S = Southampton).


In [475]:
# Changing PassengerId to string
Titanic['PassengerId'] = Titanic['PassengerId'].astype(str)

##### Code Description
---
- This code changes the data type of the **PassengerId** column to a string (object) type, preserving leading zeros and treating it as a categorical variable rather than a numerical one, preserving any leading zeros and avoiding mathematical operations that are unnecessary for identifiers.
---

In [476]:
Titanic.dtypes

PassengerId      object
Survived          int64
Pclass         category
Name           category
Sex            category
Age             float64
SibSp             int64
Parch             int64
Ticket         category
Fare            float64
Cabin          category
Embarked       category
dtype: object

#### Check for Missing Values

In [477]:
Titanic.isnull().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

##### Code description
---

**`Titanic`**: The dataset containing information about the Titanic passengers.

**`.isnull()`**: This function checks for missing values, returning `True` for missing (null) entries and `False` for present values.

**`.sum()`**: This function counts the number of `True` values (missing entries) in each column, effectively tallying the missing values.

--- 
##### Output Analysis

The columns **PassengerId**, **Survived**, **Pclass**, **Name**, **Sex**, **SibSp**, and **Ticket** have no missing values, indicating complete data.

In contrast, the **Age**, **Cabin**, and **Embarked** columns contain missing values, which need to be addressed for thorough analysis.





##### Handling  Missing Values

In [478]:
# Fill missing values in the Age column with the mean

# Calculate the mean age
mean_age = Titanic['Age'].mean()
Titanic['Age'] = Titanic['Age'].fillna(mean_age)

##### Code description
---
-The code calculates the mean age of passengers using the `mean()` function on the **Age** column. It then uses `fillna()`It then replaces any missing values in the **Age** and assigning the result back to the Age column directly.

-The mean is used for the **Age** column because it provides a central value that represents the average age of passengers, making it a suitable choice for filling missing values in a numerical dataset. This approach minimizes distortion of the overall age distribution compared to filling with a value that may be too high or too low.

---

In [479]:
# Fill missing values in the Embarked column with the mode
mode_embarked = Titanic['Embarked'].mode()[0]  # Get the first mode if there are multiple
Titanic['Embarked'] = Titanic['Embarked'].fillna(mode_embarked)

##### Code description
---
- The code computes the mode of the **Embarked** column using the `mode()` function, which identifies the most frequently occurring embarkation point. It then uses fillna() to replace any missing values in the Embarked column with this mode value and assigns the result back to the original column in the DataFrame, effectively updating it without needing inplace=True.

- The mode is used for the **Embarked** column is catagorical data because it reflects the most common embarkation point among passengers. This method preserves the dataset's overall distribution and ensures that the imputed values accurately represent the majority experience of the passengers.
---


##### Conditions consider Filling missing values of `Cabin column`
-Filling missing values for the `Cabin column` in the Titanic dataset can be approached in a few ways, but given that a significant portion (687 out of 891) is missing, it's important to choose a method that maintains the integrity of the  analysis. 

---

##### Imputation Based on Other Features

In [480]:
# Fill missing values in the Cabin column based on Pclass
for pclass in Titanic['Pclass'].unique():
    most_common_cabin = Titanic[Titanic['Pclass'] == pclass]['Cabin'].mode()[0]
    Titanic.loc[(Titanic['Pclass'] == pclass) & (Titanic['Cabin'].isnull()), 'Cabin'] = most_common_cabin

##### Code Explanation:
- **Loop Through Unique Pclass Values**: The `for` loop iterates over each unique value in the **Pclass** column. This allows us to analyze the data for each class of passengers separately.
  
- **Calculate Most Common Cabin**: 
  - `most_common_cabin = Titanic[Titanic['Pclass'] == pclass]['Cabin'].mode()[0]`:
    - This line filters the Titanic DataFrame to include only the rows where **Pclass** equals the current `pclass` in the loop.
    - It then computes the mode (most frequently occurring value) of the **Cabin** column for that specific **Pclass**.
    - `[0]` retrieves the first mode in case there are multiple modes.

- **Fill Missing Values**:
  - `Titanic.loc[(Titanic['Pclass'] == pclass) & (Titanic['Cabin'].isnull()), 'Cabin'] = most_common_cabin`:
    - This line uses the `.loc` accessor to target rows in the DataFrame where the **Pclass** matches the current `pclass` and the **Cabin** value is missing (is null).
    - It assigns the previously calculated `most_common_cabin` to those missing **Cabin** entries.

### Summary:
This approach ensures that missing cabin values are filled with the most common cabin assignment specific to each passenger class, improving the accuracy and relevance of the data imputation process. By leveraging the known relationships between **Pclass** and cabin assignments, this method enhances the integrity of the dataset.

###  Handling Outliers by lower bound and upper bound

In [481]:
# Select numerical columns for outlier detection
numerical_cols = ['Age', 'Fare', 'SibSp', 'Parch']

# Define the Z-score threshold for outliers
threshold = 3

# Calculate Z-scores and count outliers for each column
outlier_counts = {col: (abs((Titanic[col] - Titanic[col].mean()) / Titanic[col].std()) > threshold).sum() for col in numerical_cols}

# Display the count of outliers for each column
print("Count of outliers based on Z-score method:")
print(outlier_counts)


Count of outliers based on Z-score method:
{'Age': 7, 'Fare': 20, 'SibSp': 30, 'Parch': 15}


##### Code description
---
- The code selects the numerical columns **Age**, **Fare**, **SibSp**, and **Parch** for outlier detection. It defines a Z-score threshold of 3, indicating that values beyond this threshold will be considered outliers.

- The code then calculates the Z-scores for each numerical column by subtracting the mean and dividing by the standard deviation. It counts the number of outliers in each column and stores the counts in the `outlier_counts` dictionary. This dictionary provides a clear summary of the outlier counts for each specified numerical feature, making it easy to identify which variables may have extreme values.
---


##### Formatted_output

In [482]:
# Sample outlier counts
outlier_counts = {'Age': 7, 'Fare': 20, 'SibSp': 30, 'Parch': 15}

# Display counts in a formatted way
for key, value in outlier_counts.items():
    print(f"{key} {value}")


Age 7
Fare 20
SibSp 30
Parch 15


##### Handling Outliers by lower bound and upper bound

In [483]:
# Function to identify outliers using IQR
def identify_outliers_iqr(df, column):
    Q1 = df[column].quantile(0.25)
    Q3 = df[column].quantile(0.75)
    IQR = Q3 - Q1
    
    lower_bound = Q1 - 1.5 * IQR
    upper_bound = Q3 + 1.5 * IQR
    
    return lower_bound, upper_bound

# Identify outliers for 'Age', 'SibSp', 'Fare', and 'Parch'
print("{:<10} {:<20} {:<20} {:<20}".format("Column", "Lower Bound", "Upper Bound", "Outlier Count"))
print("=" * 70)

for col in ['Age', 'SibSp', 'Fare', 'Parch']:
    lower, upper = identify_outliers_iqr(Titanic, col)
    
    # Count outliers
    outlier_count = Titanic[(Titanic[col] < lower) | (Titanic[col] > upper)].shape[0]
    
    # Print results in neat format
    print("{:<10} {:<20} {:<20} {:<20}".format(col, round(lower, 2), round(upper, 2), outlier_count))

Column     Lower Bound          Upper Bound          Outlier Count       
Age        2.5                  54.5                 66                  
SibSp      -1.5                 2.5                  46                  
Fare       -26.72               65.63                116                 
Parch      0.0                  0.0                  213                 


##### Explanation of the Code
---
- Function Definition: The identify_outliers_iqr function calculates the lower and upper bounds for a specified column using the IQR method.
- Loop Through Columns: The code iterates through the relevant numerical columns (Age, SibSp,Parch, Fare).
-Print Bounds: For each column, it prints the calculated lower and upper bounds.
-Identify and Print Outliers: It identifies outliers and prints the count and values of outliers for each column.

---

##### Capping Outliers in a DataFrame Using the Interquartile Range (IQR) Method

In [484]:
# Function to cap outliers using IQR
def cap_outliers_iqr(df, column):
    Q1 = df[column].quantile(0.25)
    Q3 = df[column].quantile(0.75)
    IQR = Q3 - Q1
    
    lower_bound = Q1 - 1.5 * IQR
    upper_bound = Q3 + 1.5 * IQR
    
    # Capping the outliers
    df[column] = df[column].clip(lower=lower_bound, upper=upper_bound)

# List of columns to cap
columns_to_cap = ['Age', 'SibSp', 'Fare', 'Parch']

# Apply capping for each specified column
for col in columns_to_cap:
    cap_outliers_iqr(Titanic, col)


##### Explanation of the Code

- The function `cap_outliers_iqr` is designed to cap outliers in a specified DataFrame column.
- It calculates the first quartile (Q1) and third quartile (Q3) to determine the interquartile range (IQR).
- Lower and upper bounds for outliers are set using \( Q1 - 1.5 \times \text{IQR} \) and \( Q3 + 1.5 \times \text{IQR} \).
- The function uses the `clip()` method to cap values that fall below or above these bounds.
- A list of columns ('Age', 'SibSp', 'Fare', 'Parch') is specified, and the capping function is applied to each of them.

##### Counting Outliers After Capping Using IQR Method

In [485]:
for col in columns_to_cap:
    lower, upper = identify_outliers_iqr(Titanic, col)
    
    # Count outliers after capping
    outlier_count_after = Titanic[(Titanic[col] < lower) | (Titanic[col] > upper)].shape[0]
    
    # Print results for outliers after capping
    print("{:<10}  {:<20}".format(col, outlier_count_after))

Age         0                   
SibSp       0                   
Fare        0                   
Parch       0                   


### Code Explanation

- The loop iterates over each column specified in `columns_to_cap`.
- For each column, it calls the `identify_outliers_iqr` function to retrieve the lower and upper bounds for outliers.
- It counts the number of outliers that remain after capping by checking which values fall outside the determined bounds.
- Finally, it prints the number of outliers that are present in each column after the capping operation.

##### Lets check before performing any anaylis

In [486]:
# Display the updated DataFrame info
Titanic.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype   
---  ------       --------------  -----   
 0   PassengerId  891 non-null    object  
 1   Survived     891 non-null    int64   
 2   Pclass       891 non-null    category
 3   Name         891 non-null    category
 4   Sex          891 non-null    category
 5   Age          891 non-null    float64 
 6   SibSp        891 non-null    float64 
 7   Parch        891 non-null    int64   
 8   Ticket       891 non-null    category
 9   Fare         891 non-null    float64 
 10  Cabin        891 non-null    category
 11  Embarked     891 non-null    category
dtypes: category(6), float64(3), int64(2), object(1)
memory usage: 116.1+ KB


##### Rearranging Titanic DataFrame Columns for Improved Clarity

In [487]:
# Rearranging the columns as per your specified order
new_order = ['PassengerId', 'Name', 'Sex', 'Age', 'Pclass', 'SibSp', 'Parch', 'Embarked','Ticket','Fare', 'Cabin']

# Reindex the DataFrame in place
Titanic = Titanic.reindex(columns=new_order)

# Display the rearranged DataFrame structure
Titanic.head()


PassengerId                                               Name     Sex  \
0           1                            Braund, Mr. Owen Harris    male   
1           2  Cumings, Mrs. John Bradley (Florence Briggs Th...  female   
2           3                             Heikkinen, Miss. Laina  female   
3           4       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female   
4           5                           Allen, Mr. William Henry    male   

    Age Pclass  SibSp  Parch Embarked            Ticket     Fare Cabin  
0  22.0      3    1.0      0        S         A/5 21171   7.2500    G6  
1  38.0      1    1.0      0        C          PC 17599  65.6344   C85  
2  26.0      3    0.0      0        S  STON/O2. 3101282   7.9250    G6  
3  35.0      1    1.0      0        S            113803  53.1000  C123  
4  35.0      3    0.0      0        S            373450   8.0500    G6

### Exploratory Data Analysis (EDA)

##### Summary Stastics

In [488]:
# Get statistical summary of numerical columns
numerical_summary = Titanic.describe()
numerical_summary

Age       SibSp  Parch        Fare
count  891.000000  891.000000  891.0  891.000000
mean    29.376817    0.426487    0.0   24.046813
std     12.062035    0.708246    0.0   20.481625
min      2.500000    0.000000    0.0    0.000000
25%     22.000000    0.000000    0.0    7.910400
50%     29.699118    0.000000    0.0   14.454200
75%     35.000000    1.000000    0.0   31.000000
max     54.500000    2.500000    0.0   65.634400

In [489]:
### for better clarirty
numerical_summary.transpose() # Transpose for easier reading

count       mean        std  min      25%        50%   75%      max
Age    891.0  29.376817  12.062035  2.5  22.0000  29.699118  35.0  54.5000
SibSp  891.0   0.426487   0.708246  0.0   0.0000   0.000000   1.0   2.5000
Parch  891.0   0.000000   0.000000  0.0   0.0000   0.000000   0.0   0.0000
Fare   891.0  24.046813  20.481625  0.0   7.9104  14.454200  31.0  65.6344

In [490]:
# Summary statistics for categorical columns
categorical_summary =Titanic.describe(include='category')
print("\nCategorical Summary Statistics:")
categorical_summary


Categorical Summary Statistics:


Name   Sex  Pclass Embarked Ticket Cabin
count                   891   891     891      891    891   891
unique                  891     2       3        3    681   147
top     Abbing, Mr. Anthony  male       3        S   1601    G6
freq                      1   577     491      646      7   483

In [491]:
# Summary statistics for object columns (includes names, tickets, etc.)
object_summary = Titanic.describe(include='object')
print("\nObject Summary Statistics:")
print(object_summary)


Object Summary Statistics:
       PassengerId
count          891
unique         891
top              1
freq             1
